### DISCUSSION TITLE
How to create "fake" client stub?

### DISCUSSION THREAD

**Discussion by lhluo** [grpc/grpc-kotlin#357] [Q&A]
Hi team, new to grpc-kotlin so would appreciate any help!

I'm going through the gRPC [kotlin tutorial here](https://grpc.io/docs/languages/kotlin/basics/#client). It seems easy to create a client stub and call it like follows:
```
val channel = ManagedChannelBuilder.forAddress("localhost", 8980).usePlaintext().build()
val stub = RouteGuideCoroutineStub(channel)
```

However, I can't figure out how to create a "fake" stub for testing / running locally. Is there a way to create a fake `RouteGuideCoroutineStub` object? Or a way of binding this stub to call a fake in-process server?

I see in the non-kt offering of grpc that there exists a function `RouteGuideCoroutine.newFutureStub` that would let us instantiate a new stub. Does something like this exist for grpc Kt?

fyi (also posted this question to [stack overflow](https://stackoverflow.com/questions/73692988/grpc-kotlin-fake-client-stub))

**Comment by lowasser** [comment]
Hi! In future, please only post your question to one place.

The recommended practice for testing is to use a fake in-process server, for which you can use [the same API as Java](https://github.com/grpc/grpc-java/blob/master/examples/src/test/java/io/grpc/examples/helloworld/HelloWorldServerTest.java).

To instantiate a new stub, use the constructor; you can see an example [here](https://grpc.io/docs/languages/kotlin/quickstart/#update-the-client).

**Comment by lhluo** [comment]
@lowasser thanks, can you provide an example of how to use the java api to create an instance of `RouteGuideCoroutineStub`? Like I mentioned, I'd like to use this outside of tests in a local instance. My code depends on the type `RouteGuideCoroutineStub`, but it's not an interface I can override to provide fake data.

**Comment by lowasser** [comment]
You don't use the Java API to create an instance of `RouteGuideCoroutineStub`, you create a custom instance of the server (possibly with fake data), and then use the Java API to install it.  Something like

```
val uniqueName = InProcessServerBuilder.generateName()
val server = InProcessServerBuilder.forName(uniqueName)
       .directExecutor() // directExecutor is fine for unit tests
       .addService(/** some custom instance of RouteGuideCoroutineImpl */)
       .build().start()
val channel = InProcessChannelBuilder.forName(uniqueName)
       .directExecutor()
       .build()
val stub = RouteGuideCoroutineStub(channel)
```

**Comment by jamesward** [comment]
The example test might also be helpful: https://github.com/grpc/grpc-kotlin/blob/master/examples/server/src/test/kotlin/io/grpc/examples/routeguide/RouteGuideServerTest.kt
